File: R/demGpCov2D.R
demGpCov2D <- function(ind=c(1,2), bw=FALSE) {
  library(fields)
  graphics.off() ## kill all devices

  source('~/mlprojects/gp/R/gaussSamp.R')
  source('~/mlprojects/gp/R/demGpSample.R')
  K = K[ind, ind]
  f = f[ind]
  x = x[ind]
  print(K)

  dev.new()
  lo = layout(matrix(c(1:4),2,2,byrow=T)); layout.show(lo)

#   dev.new();
  basePlot(K, ind)

#   dev.new() #figure(2)
  basePlot(K, ind)
  cont2 = lines(c(f[1],f[1]), c(-1,1), col='green')

#   dev.new() #figure(3)
  basePlot(K, ind)
  cont2 = lines(c(f[1],f[1]), c(-1,1), col='green')

  ## Compute conditional mean and variance
  f2Mean = K[1, 2]/K[1,1]*f[1]
  f2Var = K[2, 2] - K[1, 2]/K[1, 1]*K[1, 2]
  yval = as.matrix(seq(-1, 1, length=200))
  pdfVal = 1/sqrt(2*pi*f2Var)*exp(-0.5*(yval-f2Mean)*(yval-f2Mean)/f2Var)
  pdf = lines(pdfVal*0.25, yval, col='red')
#   set(pdf, 'linewidth', conditionalSize)
#   set(pdf, 'linestyle', conditionalLineStyle, 'color', conditionalLineColour)

  if (bw )
    app = 'bw'
  else 
    app = ''

  for (figNo in 2:4) {
    dev.set(figNo) #figure(figNo)
    if (exists('printDiagram') && exists('var') && exists(printDiagram))
      printPlot(paste('demGpCov2D',as.character(ind[1]),'_',as.character(ind[2]),'_', 
	as.character(figNo),app,sep=""), '../tex/diagrams', '../html')
  }
}

basePlot <- function (K, ind) {
  # BASEPLOT Plot the contour of the covariance.
  # DESC creates the basic plot.

  eigVecs = eigen(K)
  U = eigVecs$vectors[,dim(eigVecs$vectors)[2]:1]; V = rev(eigVecs$values) ## in incr. order
  V[V<0] = as.complex(V[V<0])
  r = Re(sqrt(V))
  theta = seq(0, 2*pi, length=200)
  xy = cbind(r[1]*sin(theta), r[2]*cos(theta))%*%U
  plot(0, type = "n", xlim=c(min(xy[,1]), max(xy[,1])),
    ylim=c(min(xy[,2]), max(xy[,2])), xlab='', ylab='') ## lines only works on existing plots
  cont = lines(xy[, 1], xy[, 2], col='blue')

#   t = text(0.5, -0.2, ['f_{' num2str(ind(1)) '}']);
#   t =[t text(-0.2, -0.5, ['f_{' num2str(ind(2)) '}'])];

# 
#   set(gca, 'xtick', [-1  0  1])
#   set(gca, 'ytick', [-1 0 1])
# 
#   set(t, 'fontname', 'times')
#   set(t, 'fontsize', 24)
#   set(t, 'fontangle', 'italic')
# 
#   zeroAxes(gca, 0.025, 18, 'times')
  abline(v = 0, h = 0) ## axes and co-centric circles
  int = (max(xy[,1])-min(xy[,1]))/100
  lines(int*cos(seq(0, 2*pi, l=100)), int*sin(seq(0, 2*pi, l=100)))

#   ax = gca

#   return list(ax, cont, t)
}

File: gpmgp/demGpCov2D.m
function demGpCov2D(ind,bw)

% DEMGPCOV2D Simple demonstration of sampling from a covariance function.
% FORMAT
% DESC shows two dimensions of the covariance matrix giving the joint
% distribution and the conditional. DEMGPSAMPLE is run first to set the
% covariance matrix.
% ARG index : the indices of the two elements from the covariance matrix
% for which the joint distribution should be displayed.
% ARG bw : whether or not to prepare the plot with a black and white
% output (default false).
%
% SEEALSO : demCovFuncSample, demGPSample
% COPYRIGHT : Neil D. Lawrence, 2006, 2008

% GP

if nargin < 2
  bw = false;
end
  
global printDiagram
if bw == true
  contourColour = [0 0 0];
  contourLineStyle = '--';
  conditionalLineStyle = '-.';
  conditionalLineColour = [0 0 0];
  conditionalSize = 4;
  fixedLineStyle = '-';
  fixedLineColour = [0 0 0];
  fixedSize = 4;
else
  contourColour = [0 0 1];
  contourLineStyle = '-';
  conditionalLineStyle = '-';
  conditionalLineColour = [1 0 0];
  conditionalSize = 4;
  fixedLineStyle = '-';
  fixedLineColour = [0 1 0];
  fixedSize = 4;
end
contourSize = 4;
if nargin < 1
  ind = [1 2];
end
load demGpSample
K = K(ind, ind);
f = f(ind);
x = x(ind);
disp(K);
figure(1)
clf
[ax, cont, t] = basePlot(K, ind);
set(cont, 'linewidth', contourSize);
set(cont, 'linestyle', contourLineStyle, 'color', contourColour);
figure(2)
clf
[ax, cont, t] = basePlot(K, ind);
set(cont, 'linewidth', contourSize);
set(cont, 'linestyle', contourLineStyle, 'color', contourColour);

cont2 = line([f(1) f(1)], [-1 1]);
set(cont2, 'linewidth', fixedSize)
set(cont2, 'linestyle', fixedLineStyle, 'color', fixedLineColour)

figure(3)
clf
[ax, cont, t] = basePlot(K, ind);
set(cont, 'linewidth', contourSize);
set(cont, 'linestyle', contourLineStyle, 'color', contourColour);

cont2 = line([f(1) f(1)], [-1 1]);
set(cont2, 'linewidth', fixedSize)
set(cont2, 'linestyle', fixedLineStyle, 'color', fixedLineColour)

% Compute conditional mean and variance
f2Mean = K(1, 2)/K(1, 1)*f(1);
f2Var = K(2, 2) - K(1, 2)/K(1, 1)*K(1, 2);
yval = linspace(-1, 1, 200);
pdfVal = 1/sqrt(2*pi*f2Var)*exp(-0.5*(yval-f2Mean).*(yval-f2Mean)/f2Var);
pdf = line(pdfVal*0.25, yval);
set(pdf, 'linewidth', conditionalSize)
set(pdf, 'linestyle', conditionalLineStyle, 'color', conditionalLineColour)

if bw 
  app = 'bw';
else 
  app = '';
end
for figNo = 1:3
  figure(figNo)
  if exist('printDiagram', 'var') & printDiagram
    printPlot(['demGpCov2D' num2str(ind(1)) '_' num2str(ind(2)) '_' ...
               num2str(figNo) app], '../tex/diagrams', '../html');
  end
end

function [ax, cont, t] = basePlot(K, ind)

% BASEPLOT Plot the contour of the covariance.
% FORMAT
% DESC creates the basic plot.
% 

[U, V] = eig(K);
r = sqrt(diag(V));
theta = linspace(0, 2*pi, 200)';
xy = [r(1)*sin(theta) r(2)*cos(theta)]*U;
cont = line(xy(:, 1), xy(:, 2));


t = text(0.5, -0.2, ['f_{' num2str(ind(1)) '}']);
t =[t text(-0.2, -0.5, ['f_{' num2str(ind(2)) '}'])];


set(gca, 'xtick', [-1  0  1])
set(gca, 'ytick', [-1 0 1])

set(t, 'fontname', 'times')
set(t, 'fontsize', 24)
set(t, 'fontangle', 'italic')

zeroAxes(gca, 0.025, 18, 'times')

ax = gca;
